src/main.ts

```diff
@@ -156,33 +185,57 @@ ArgoCD Diff for commit [\`${shortCommitSha}\`](${commitLink})
   }
 }
 
+async function asyncForEach<T>(
+  array: T[],
+  callback: (item: T, i: number, arr: T[]) => Promise<void>
+): Promise<void> {
+  for (let index = 0; index < array.length; index++) {
+    await callback(array[index], index, array);
+  }
+}
+
 async function run(): Promise<void> {
   const argocd = await setupArgoCDCommand();
   const apps = await getApps();
   core.info(`Found apps: ${apps.map(a => a.metadata.name).join(', ')}`);
 
-  const diffPromises = apps.map(async app => {
+  const diffs: Diff[] = [];
+
+  await asyncForEach(apps, async app => {
     try {
       const command = `app diff ${app.metadata.name} --local=${app.spec.source.path}`;
-      if (app.spec.source.helm) {
-        const pwd = await execCommand(`pwd`);
-        core.info(`pwd: ${pwd.stdout}`);
-        const output = await execCommand(`cd ${app.spec.source.path} && ls -al`);
-        core.info(`output: ${output.stdout}`);
-      }
-      const res = await argocd(command);
-      core.info(`Running: argocd ${command}`);
-      core.info(`stdout: ${res.stdout}`);
-      core.info(`stdout: ${res.stderr}`);
-      if (res.stdout) {
-        return { app, diff: res.stdout };
+      try {
+        core.info(`Running: argocd ${command}`);
+        // ArgoCD app diff will exit 1 if there is a diff, so always catch,
+        // and then consider it a success if there's a diff in stdout
+        // https://github.com/argoproj/argo-cd/issues/3588
+        await argocd(command);
+      } catch (e) {
+        const res = e as ExecResult;
+        core.info(`stdout: ${res.stdout}`);
+        core.info(`stderr: ${res.stderr}`);
+        if (res.stdout) {
+          diffs.push({ app, diff: res.stdout });
+        } else {
+          diffs.push({
+            app,
+            diff: '',
+            error: `
+stderr: ${res.stderr}
+err: ${JSON.stringify(res.err)}
```

Comment:
It'd be nice to split the error and JSON into separate blocks:

## ArgoCD Diff for commit [`f011134`](https://github.com/quizlet/quizlet-infrastructure/pull/416/commits/f0111346dc6a0007dfc30eb35b6a17bc7cf9ce05)
     

[`webapp-dev-testing-datadog`](https://argocd.qzlt.io/applications/webapp-dev-testing-datadog)  [Sync: ✅, Diff: 🛑]

```
time="2020-08-27T22:06:57Z" level=info msg="helm template . --name-template datadog --namespace monitoring --values values/common.yaml --values values/dev.yaml --values values/dev-testing.yaml --include-crds" dir=apps/datadog execID=Hbz6e
time="2020-08-27T22:06:57Z" level=error msg="`helm template . --name-template datadog --namespace monitoring --values values/common.yaml --values values/dev.yaml --values values/dev-testing.yaml --include-crds` failed exit status 1: Error: found in Chart.yaml, but missing in charts/ directory: datadog-foobar" execID=Hbz6e
time="2020-08-27T22:06:57Z" level=info msg="helm dependency build" dir=apps/datadog execID=vN98O
time="2020-08-27T22:07:00Z" level=error msg="`helm dependency build` failed exit status 1: Error: could not find : chart datadog-foobar not found in https://kubernetes-charts.storage.googleapis.com/" execID=vN98O
time="2020-08-27T22:07:00Z" level=fatal msg="`helm dependency build` failed exit status 1: Error: could not find : chart datadog-foobar not found in https://kubernetes-charts.storage.googleapis.com/"
```

```json
{"killed":false,"code":1,"signal":null,"cmd":"bin/argo app diff webapp-dev-testing-datadog --local=apps/datadog --auth-token=*** --server=argocd.qzlt.io --grpc-web"}
```

---
   
[`webapp-prod-datadog`](https://argocd.qzlt.io/applications/webapp-prod-datadog) [Sync: ✅, Diff: 🛑]

```helm
time="2020-08-27T22:07:17Z" level=info msg="helm template . --name-template datadog --namespace monitoring --values values/common.yaml --values values/prod.yaml --include-crds" dir=apps/datadog execID=bW5cW
time="2020-08-27T22:07:18Z" level=error msg="`helm template . --name-template datadog --namespace monitoring --values values/common.yaml --values values/prod.yaml --include-crds` failed exit status 1: Error: found in Chart.yaml, but missing in charts/ directory: datadog-foobar" execID=bW5cW
time="2020-08-27T22:07:18Z" level=info msg="helm dependency build" dir=apps/datadog execID=6c2jP
time="2020-08-27T22:07:20Z" level=error msg="`helm dependency build` failed exit status 1: Error: could not find : chart datadog-foobar not found in https://kubernetes-charts.storage.googleapis.com/" execID=6c2jP
time="2020-08-27T22:07:20Z" level=fatal msg="`helm dependency build` failed exit status 1: Error: could not find : chart datadog-foobar not found in https://kubernetes-charts.storage.googleapis.com/"
```

```json
{"killed":false,"code":1,"signal":null,"cmd":"bin/argo app diff webapp-prod-datadog --local=apps/datadog --auth-token=*** --server=argocd.qzlt.io --grpc-web"}
```

---

[`webapp-prod-webapp`](https://argocd.qzlt.io/applications/webapp-prod-webapp) [Sync: ✅, Diff: ✅]

<details>

```diff
===== argoproj.io/CronWorkflow hhvm/cron-send-premium-content-remarketing ======
84c84
<         image: gcr.io/compact-sylph-785/hhvm-web:09179558c1242f0cf24b448086bbaf8a0d2ac0f9
---
>         image: gcr.io/compact-sylph-785/hhvm-web:4896afc24850632f9c278a28959e3a5a550642c1
===== argoproj.io/CronWorkflow hhvm/cron-update-exchange-rates ======
84c84
<         image: gcr.io/compact-sylph-785/hhvm-web:09179558c1242f0cf24b448086bbaf8a0d2ac0f9
---
>         image: gcr.io/compact-sylph-785/hhvm-web:4896afc24850632f9c278a28959e3a5a550642c1
===== argoproj.io/CronWorkflow hhvm/cron-update-set-page-urls ======
84c84
<         image: gcr.io/compact-sylph-785/hhvm-web:09179558c1242f0cf24b448086bbaf8a0d2ac0f9
---
>         image: gcr.io/compact-sylph-785/hhvm-web:4896afc24850632f9c278a28959e3a5a550642c1
===== argoproj.io/CronWorkflow hhvm/cron-agora-daily-purchasebot ======
84c84
<         image: gcr.io/compact-sylph-785/hhvm-web:09179558c1242f0cf24b448086bbaf8a0d2ac0f9
---
>         image: gcr.io/compact-sylph-785/hhvm-web:4896afc24850632f9c278a28959e3a5a550642c1
===== argoproj.io/CronWorkflow hhvm/cron-subs-renewal-reminder-email-recurly ======
84c84
<         image: gcr.io/compact-sylph-785/hhvm-web:09179558c1242f0cf24b448086bbaf8a0d2ac0f9
---
>         image: gcr.io/compact-sylph-785/hhvm-web:4896afc24850632f9c278a28959e3a5a550642c1
===== argoproj.io/Rollout hhvm/web-search ======
39c39
<         image: gcr.io/compact-sylph-785/hhvm-web:09179558c1242f0cf24b448086bbaf8a0d2ac0f9
---
>         image: gcr.io/compact-sylph-785/hhvm-web:4896afc24850632f9c278a28959e3a5a550642c1
===== argoproj.io/CronWorkflow hhvm/cron-android-plus-winback-2020--email ======
84c84
<         image: gcr.io/compact-sylph-785/hhvm-web:09179558c1242f0cf24b448086bbaf8a0d2ac0f9
---
>         image: gcr.io/compact-sylph-785/hhvm-web:4896afc24850632f9c278a28959e3a5a550642c1
===== argoproj.io/CronWorkflow hhvm/cron-gen-dev-db-daily ======
84c84
<         image: gcr.io/compact-sylph-785/hhvm-web:09179558c1242f0cf24b448086bbaf8a0d2ac0f9
---
>         image: gcr.io/compact-sylph-785/hhvm-web:4896afc24850632f9c278a28959e3a5a550642c1
===== argoproj.io/CronWorkflow hhvm/cron-multiplayer-server-old-game-cleanup ======
84c84
<         image: gcr.io/compact-sylph-785/hhvm-web:09179558c1242f0cf24b448086bbaf8a0d2ac0f9
---
>         image: gcr.io/compact-sylph-785/hhvm-web:4896afc24850632f9c278a28959e3a5a550642c1
===== argoproj.io/CronWorkflow hhvm/cron-geolocate-users ======
84c84
<         image: gcr.io/compact-sylph-785/hhvm-web:09179558c1242f0cf24b448086bbaf8a0d2ac0f9
---
>         image: gcr.io/compact-sylph-785/hhvm-web:4896afc24850632f9c278a28959e3a5a550642c1
===== argoproj.io/CronWorkflow hhvm/cron-cleanup-client-id-mappings ======
84c84
<         image: gcr.io/compact-sylph-785/hhvm-web:09179558c1242f0cf24b448086bbaf8a0d2ac0f9
---
>         image: gcr.io/compact-sylph-785/hhvm-web:4896afc24850632f9c278a28959e3a5a550642c1
===== argoproj.io/CronWorkflow hhvm/cron-delete-old-carts ======
84c84
<         image: gcr.io/compact-sylph-785/hhvm-web:09179558c1242f0cf24b448086bbaf8a0d2ac0f9
---
>         image: gcr.io/compact-sylph-785/hhvm-web:4896afc24850632f9c278a28959e3a5a550642c1
===== argoproj.io/CronWorkflow hhvm/cron-detect-similar-schools ======
85c85
<         image: gcr.io/compact-sylph-785/hhvm-web:09179558c1242f0cf24b448086bbaf8a0d2ac0f9
---
>         image: gcr.io/compact-sylph-785/hhvm-web:4896afc24850632f9c278a28959e3a5a550642c1
===== argoproj.io/CronWorkflow hhvm/cron-check-spam-classifier ======
84c84
<         image: gcr.io/compact-sylph-785/hhvm-web:09179558c1242f0cf24b448086bbaf8a0d2ac0f9
---
>         image: gcr.io/compact-sylph-785/hhvm-web:4896afc24850632f9c278a28959e3a5a550642c1
===== argoproj.io/CronWorkflow hhvm/cron-localize-sendgrid-templates ======
84c84
<         image: gcr.io/compact-sylph-785/hhvm-web:09179558c1242f0cf24b448086bbaf8a0d2ac0f9
---
>         image: gcr.io/compact-sylph-785/hhvm-web:4896afc24850632f9c278a28959e3a5a550642c1
===== argoproj.io/CronWorkflow hhvm/cron-trim-web-events ======
84c84
<         image: gcr.io/compact-sylph-785/hhvm-web:09179558c1242f0cf24b448086bbaf8a0d2ac0f9
---
>         image: gcr.io/compact-sylph-785/hhvm-web:4896afc24850632f9c278a28959e3a5a550642c1
===== argoproj.io/CronWorkflow hhvm/cron-reindex-popular-courses ======
84c84
<         image: gcr.io/compact-sylph-785/hhvm-web:09179558c1242f0cf24b448086bbaf8a0d2ac0f9
---
>         image: gcr.io/compact-sylph-785/hhvm-web:4896afc24850632f9c278a28959e3a5a550642c1
===== argoproj.io/Rollout hhvm/web ======
41c41
<         image: gcr.io/compact-sylph-785/hhvm-web:09179558c1242f0cf24b448086bbaf8a0d2ac0f9
---
>         image: gcr.io/compact-sylph-785/hhvm-web:4896afc24850632f9c278a28959e3a5a550642c1
===== argoproj.io/CronWorkflow hhvm/cron-gen-dev-db-weekly ======
84c84
<         image: gcr.io/compact-sylph-785/hhvm-web:09179558c1242f0cf24b448086bbaf8a0d2ac0f9
---
>         image: gcr.io/compact-sylph-785/hhvm-web:4896afc24850632f9c278a28959e3a5a550642c1
===== argoproj.io/CronWorkflow hhvm/cron-onboarding-teacher-emails ======
84c84
<         image: gcr.io/compact-sylph-785/hhvm-web:09179558c1242f0cf24b448086bbaf8a0d2ac0f9
---
>         image: gcr.io/compact-sylph-785/hhvm-web:4896afc24850632f9c278a28959e3a5a550642c1
===== argoproj.io/CronWorkflow hhvm/cron-send-premium-rating-request ======
84c84
<         image: gcr.io/compact-sylph-785/hhvm-web:09179558c1242f0cf24b448086bbaf8a0d2ac0f9
---
>         image: gcr.io/compact-sylph-785/hhvm-web:4896afc24850632f9c278a28959e3a5a550642c1
===== argoproj.io/CronWorkflow hhvm/cron-subs-expire-lapsed-recurring ======
84c84
<         image: gcr.io/compact-sylph-785/hhvm-web:09179558c1242f0cf24b448086bbaf8a0d2ac0f9
---
>         image: gcr.io/compact-sylph-785/hhvm-web:4896afc24850632f9c278a28959e3a5a550642c1
===== argoproj.io/CronWorkflow hhvm/cron-update-folder-urls ======
85c85
<         image: gcr.io/compact-sylph-785/hhvm-web:09179558c1242f0cf24b448086bbaf8a0d2ac0f9
---
>         image: gcr.io/compact-sylph-785/hhvm-web:4896afc24850632f9c278a28959e3a5a550642c1
===== argoproj.io/CronWorkflow hhvm/cron-onboarding-student-emails ======
84c84
<         image: gcr.io/compact-sylph-785/hhvm-web:09179558c1242f0cf24b448086bbaf8a0d2ac0f9
---
>         image: gcr.io/compact-sylph-785/hhvm-web:4896afc24850632f9c278a28959e3a5a550642c1
===== argoproj.io/Rollout hhvm/web-suggestions ======
39c39
<         image: gcr.io/compact-sylph-785/hhvm-web:09179558c1242f0cf24b448086bbaf8a0d2ac0f9
---
>         image: gcr.io/compact-sylph-785/hhvm-web:4896afc24850632f9c278a28959e3a5a550642c1
===== argoproj.io/Rollout hhvm/web-worker ======
44c44
<         image: gcr.io/compact-sylph-785/hhvm-web:09179558c1242f0cf24b448086bbaf8a0d2ac0f9
---
>         image: gcr.io/compact-sylph-785/hhvm-web:4896afc24850632f9c278a28959e3a5a550642c1
===== argoproj.io/CronWorkflow hhvm/cron-send-premium-content-prep-pack-remarketing ======
84c84
<         image: gcr.io/compact-sylph-785/hhvm-web:09179558c1242f0cf24b448086bbaf8a0d2ac0f9
---
>         image: gcr.io/compact-sylph-785/hhvm-web:4896afc24850632f9c278a28959e3a5a550642c1
===== argoproj.io/CronWorkflow hhvm/cron-send-premium-content-set-to-prep-pack-remarketing ======
84c84
<         image: gcr.io/compact-sylph-785/hhvm-web:09179558c1242f0cf24b448086bbaf8a0d2ac0f9
---
>         image: gcr.io/compact-sylph-785/hhvm-web:4896afc24850632f9c278a28959e3a5a550642c1
===== argoproj.io/CronWorkflow hhvm/cron-send-renew-reminders ======
84c84
<         image: gcr.io/compact-sylph-785/hhvm-web:09179558c1242f0cf24b448086bbaf8a0d2ac0f9
---
>         image: gcr.io/compact-sylph-785/hhvm-web:4896afc24850632f9c278a28959e3a5a550642c1
===== argoproj.io/CronWorkflow hhvm/cron-temp-check-qbatch ======
85c85
<         image: gcr.io/compact-sylph-785/hhvm-web:09179558c1242f0cf24b448086bbaf8a0d2ac0f9
---
>         image: gcr.io/compact-sylph-785/hhvm-web:4896afc24850632f9c278a28959e3a5a550642c1
===== argoproj.io/CronWorkflow hhvm/cron-send-hourly-notification ======
84c84
<         image: gcr.io/compact-sylph-785/hhvm-web:09179558c1242f0cf24b448086bbaf8a0d2ac0f9
---
>         image: gcr.io/compact-sylph-785/hhvm-web:4896afc24850632f9c278a28959e3a5a550642c1
===== argoproj.io/CronWorkflow hhvm/cron-subs-renewal-reminder-email-google ======
84c84
<         image: gcr.io/compact-sylph-785/hhvm-web:09179558c1242f0cf24b448086bbaf8a0d2ac0f9
---
>         image: gcr.io/compact-sylph-785/hhvm-web:4896afc24850632f9c278a28959e3a5a550642c1
===== argoproj.io/CronWorkflow hhvm/cron-sync-recurring-subscriptions-apple ======
84c84
<         image: gcr.io/compact-sylph-785/hhvm-web:09179558c1242f0cf24b448086bbaf8a0d2ac0f9
---
>         image: gcr.io/compact-sylph-785/hhvm-web:4896afc24850632f9c278a28959e3a5a550642c1
===== argoproj.io/CronWorkflow hhvm/cron-trim-visitor-datastore-stale-rows ======
84c84
<         image: gcr.io/compact-sylph-785/hhvm-web:09179558c1242f0cf24b448086bbaf8a0d2ac0f9
---
>         image: gcr.io/compact-sylph-785/hhvm-web:4896afc24850632f9c278a28959e3a5a550642c1
===== argoproj.io/CronWorkflow hhvm/cron-agora-weekly-purchasebot ======
84c84
<         image: gcr.io/compact-sylph-785/hhvm-web:09179558c1242f0cf24b448086bbaf8a0d2ac0f9
---
>         image: gcr.io/compact-sylph-785/hhvm-web:4896afc24850632f9c278a28959e3a5a550642c1
===== argoproj.io/CronWorkflow hhvm/cron-cleanup-measure-enrollment ======
85c85
<         image: gcr.io/compact-sylph-785/hhvm-web:09179558c1242f0cf24b448086bbaf8a0d2ac0f9
---
>         image: gcr.io/compact-sylph-785/hhvm-web:4896afc24850632f9c278a28959e3a5a550642c1
===== argoproj.io/CronWorkflow hhvm/cron-generate-premium-folder-sitemap ======
84c84
<         image: gcr.io/compact-sylph-785/hhvm-web:09179558c1242f0cf24b448086bbaf8a0d2ac0f9
---
>         image: gcr.io/compact-sylph-785/hhvm-web:4896afc24850632f9c278a28959e3a5a550642c1
===== argoproj.io/CronWorkflow hhvm/cron-localize-braze-templates ======
84c84
<         image: gcr.io/compact-sylph-785/hhvm-web:09179558c1242f0cf24b448086bbaf8a0d2ac0f9
---
>         image: gcr.io/compact-sylph-785/hhvm-web:4896afc24850632f9c278a28959e3a5a550642c1
===== argoproj.io/CronWorkflow hhvm/cron-subs-renewal-reminder-email-apple ======
84c84
<         image: gcr.io/compact-sylph-785/hhvm-web:09179558c1242f0cf24b448086bbaf8a0d2ac0f9
---
>         image: gcr.io/compact-sylph-785/hhvm-web:4896afc24850632f9c278a28959e3a5a550642c1
===== argoproj.io/CronWorkflow hhvm/cron-sync-recurring-subscriptions-google ======
84c84
<         image: gcr.io/compact-sylph-785/hhvm-web:09179558c1242f0cf24b448086bbaf8a0d2ac0f9
---
>         image: gcr.io/compact-sylph-785/hhvm-web:4896afc24850632f9c278a28959e3a5a550642c1
===== argoproj.io/CronWorkflow hhvm/cron-anonymize-deleted-users ======
84c84
<         image: gcr.io/compact-sylph-785/hhvm-web:09179558c1242f0cf24b448086bbaf8a0d2ac0f9
---
>         image: gcr.io/compact-sylph-785/hhvm-web:4896afc24850632f9c278a28959e3a5a550642c1
===== argoproj.io/CronWorkflow hhvm/cron-send-premium-content-expiration-notice ======
84c84
<         image: gcr.io/compact-sylph-785/hhvm-web:09179558c1242f0cf24b448086bbaf8a0d2ac0f9
---
>         image: gcr.io/compact-sylph-785/hhvm-web:4896afc24850632f9c278a28959e3a5a550642c1

```

</details>

Reply:
I could also see an argument for hiding output behind `<details>` only when it exceeds a certain length.

Reply:
+1 will follow up with this